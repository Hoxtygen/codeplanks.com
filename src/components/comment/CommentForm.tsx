import { useFormik } from "formik";
import React from "react";
import { CommentFormProps } from "../../typedefs";

import { commentFormInitialValues } from "./formUtils/initialValues";
import { validationSchema } from "./formUtils/validationSchema";

export default function CommentForm({
  slug,
  handlePostSubmission,
  showSuccessMessage,
}: CommentFormProps) {
  const formik = useFormik({
    initialValues: commentFormInitialValues,
    validationSchema: validationSchema,
    onSubmit: (values) => {
      values.slug = slug;
      handlePostSubmission(values);
      formik.resetForm();
    },
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <div className="bg-white dark:bg-slate-800 bg-white shadow-lg rounded-lg p-8 pb-12 mb-8">
        <h3 className="text-xl mb-8 font-semibold border-b pb-4">
          Leave a Reply
        </h3>

        <div className="grid grid-cols-1 gap-4 mb-4">
          <input
            className="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
            name="name"
            onChange={formik.handleChange}
            placeholder="Name"
            type="text"
            value={formik.values.name}
            onBlur={formik.handleBlur}
            data-testid="name"
          />
          <p className="text-xs text-red-500">
            {formik.touched.name && Boolean(formik.errors.name)
              ? formik.errors.name
              : null}
          </p>

          <input
            className="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
            name="email"
            onChange={formik.handleChange}
            placeholder="Email"
            type="email"
            onBlur={formik.handleBlur}
            value={formik.values.email}
            data-testid="email"
          />
          <p className="text-xs text-red-500">
            {formik.touched.email && Boolean(formik.errors.email)
              ? formik.errors.email
              : null}
          </p>
        </div>
        <div className="grid grid-cols-1 gap-4 mb-4">
          <textarea
            className="p-4 outline-none w-full rounded-lg h-40 focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
            cols={30}
            id="comment"
            name="comment"
            placeholder="Comment"
            rows={10}
            value={formik.values.comment}
            onBlur={formik.handleBlur}
            onChange={formik.handleChange}
            data-testid="comment"
          />
          <p className="text-xs text-red-500">
            {formik.touched.comment && Boolean(formik.errors.comment)
              ? formik.errors.comment
              : null}
          </p>
        </div>

        <div className="mt-8">
          <button
            type="submit"
            className="transition duration-500 ease hover:bg-indigo-900 inline-block bg-pink-600 text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
            disabled={!(formik.isValid && formik.dirty)}
          >
            Post Comment
          </button>
          {showSuccessMessage && (
            <span className="text-xl float-right font-semibold mt-3 text-green-500">
              Comment submitted for review
            </span>
          )}
        </div>
      </div>
    </form>
  );
}
