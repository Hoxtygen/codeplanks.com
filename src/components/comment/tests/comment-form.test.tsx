import React from "react";
import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import CommentForm from "../CommentForm";
import userEvent from "@testing-library/user-event";

describe("CommentForm", () => {
  it("should render comment form", () => {
    const handleSubmit = jest.fn();
    const slug = "promises";
    render(
      <CommentForm
        showSuccessMessage={true}
        slug={slug}
        handlePostSubmission={handleSubmit}
      />
    );

    expect(screen.getByTestId("email")).toBeInTheDocument();
    expect(screen.getByTestId("email")).toMatchSnapshot();
    expect(screen.getByTestId("comment")).toBeInTheDocument();
    expect(screen.getByTestId("comment")).toMatchSnapshot();
    expect(
      screen.getByRole("button", { name: /Post comment/i })
    ).toBeInTheDocument();
    expect(screen.getByTestId("name")).toBeInTheDocument();
    expect(screen.getByTestId("name")).toMatchSnapshot();
    expect(
      screen.getByRole("button", { name: /Post comment/i })
    ).toMatchSnapshot();
  });

  it("should submit comment form", async () => {
    const handleSubmit = jest.fn();
    const user = userEvent.setup();
    const slug = "promises";
    render(
      <CommentForm
        showSuccessMessage={true}
        slug={slug}
        handlePostSubmission={handleSubmit}
      />
    );
    await userEvent.type(screen.getByTestId("name"), "Mark Anderson");
    await userEvent.type(screen.getByTestId("email"), "Anderson@example.com");
    await userEvent.type(
      screen.getByTestId("comment"),
      "Lorem ipsum dolor sit amet adipscing consectetur"
    );

    await user.click(screen.getByRole("button", { name: /Post comment/i }));
    await waitFor(() =>
      expect(handleSubmit).toHaveBeenCalledWith({
        name: "Mark Anderson",
        email: "Anderson@example.com",
        comment: "Lorem ipsum dolor sit amet adipscing consectetur",
        slug: slug,
      })
    );
  });

  it("should not submit comment when email is invalid", async () => {
    const handleSubmit = jest.fn();
    const user = userEvent.setup();
    const slug = "promises";
    render(
      <CommentForm
        showSuccessMessage={true}
        slug={slug}
        handlePostSubmission={handleSubmit}
      />
    );
    await userEvent.type(screen.getByTestId("name"), "Mark Anderson");
    await userEvent.type(screen.getByTestId("email"), "Anderson@example");
    await userEvent.type(
      screen.getByTestId("comment"),
      "Lorem ipsum dolor sit amet adipscing consectetur"
    );

    await user.click(screen.getByRole("button", { name: /Post comment/i }));
    expect(screen.getByRole("button")).toBeDisabled();
    await waitFor(() => expect(handleSubmit).not.toHaveBeenCalled());
  });

  it("should not submit comment when name is less than 2 characters", async () => {
    const handleSubmit = jest.fn();
    const user = userEvent.setup();
    const slug = "promises";
    render(
      <CommentForm
        showSuccessMessage={true}
        slug={slug}
        handlePostSubmission={handleSubmit}
      />
    );
    await userEvent.type(screen.getByTestId("name"), "M");
    await userEvent.type(screen.getByTestId("email"), "Anderson@example.com");
    await userEvent.type(
      screen.getByTestId("comment"),
      "Lorem ipsum dolor sit amet adipscing consectetur"
    );

    await user.click(screen.getByRole("button", { name: /Post comment/i }));
    expect(screen.getByRole("button")).toBeDisabled();
    await waitFor(() => expect(handleSubmit).not.toHaveBeenCalled());
  });

  it("should not submit comment when comment is less than 5 characters", async () => {
    const handleSubmit = jest.fn();
    const user = userEvent.setup();
    const slug = "promises";
    render(
      <CommentForm
        showSuccessMessage={true}
        slug={slug}
        handlePostSubmission={handleSubmit}
      />
    );
    await userEvent.type(screen.getByTestId("name"), "Mark Anderson");
    await userEvent.type(screen.getByTestId("email"), "Anderson@example.com");
    await userEvent.type(screen.getByTestId("comment"), "Lore");

    await user.click(screen.getByRole("button", { name: /Post comment/i }));
    expect(screen.getByRole("button")).toBeDisabled();
    await waitFor(() => expect(handleSubmit).not.toHaveBeenCalled());
  });
});
