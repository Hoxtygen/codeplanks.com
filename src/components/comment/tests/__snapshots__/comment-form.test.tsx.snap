// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentForm should render comment form 1`] = `
<input
  class="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
  data-testid="email"
  name="email"
  placeholder="Email"
  type="email"
  value=""
/>
`;

exports[`CommentForm should render comment form 2`] = `
<textarea
  class="p-4 outline-none w-full rounded-lg h-40 focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
  cols="30"
  data-testid="comment"
  id="comment"
  name="comment"
  placeholder="Comment"
  rows="10"
/>
`;

exports[`CommentForm should render comment form 3`] = `
<input
  class="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
  data-testid="name"
  name="name"
  placeholder="Name"
  type="text"
  value=""
/>
`;

exports[`CommentForm should render comment form 4`] = `
<button
  class="transition duration-500 ease hover:bg-indigo-900 inline-block bg-pink-600 text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
  disabled=""
  type="submit"
>
  Post Comment
</button>
`;
